### ISSUE TITLE
Error 400 on Creating CosmosDB with Private Link

### ISSUE THREAD

**Issue by akamalov** [terraform-providers/terraform-provider-azurerm#5864]
Greetings,

*Environment:*
```
Terraform v0.12.8
AzureRM: 1.39.0
```

Trying to deploy CosmosDB on Azure with Private Link. Followed Microsoft guidelines on disabling network policies for private link -  https://docs.microsoft.com/en-us/azure/private-link/disable-private-endpoint-network-policy. Getting an error 400, which is very generic in nature.

*Code:*

**File: main.tf**
```
resource "random_id" "id" {
  byte_length = 4
}

#Create resource group
resource "azurerm_resource_group" "resource_group" {
  name     = local.resource_group_name
  location = var.primary_location
  tags = var.tags
}

# Create Azure Cosmos Account 
resource "azurerm_cosmosdb_account" "account" {
  name                = lower("${var.account_kind}-${local.outapplicationname}-${random_id.id.hex}")
  location            = azurerm_resource_group.resource_group.location
  resource_group_name = azurerm_resource_group.resource_group.name
  # tags = var.tags
  offer_type          = "Standard"
  kind                = var.account_kind
  enable_automatic_failover = var.is_env_prod ? true : false
  is_virtual_network_filter_enabled = true # Default block all traffic with firewall
  ip_range_filter = xxx.xxx.xxx.xxx,xxx.xxx.xxx.xxx,xxx.xxx.xxx.xxx" # Azure portal ips

  capabilities {
    name = var.account_capabilities
  }

  consistency_policy {
    consistency_level       = "${var.consistency_level}"
    max_interval_in_seconds = 310
    max_staleness_prefix    = 101000
  }

   dynamic "geo_location" {
    for_each = local.region_list
    content {
      location = geo_location.value
      # failover_priority = length (local.region_list) - 1
      failover_priority = index (local.region_list, "${geo_location.value}")
    }
  }

}

# Create private endpoints. If Production environment, in both primary and secondary regions. Non-prod environment, only in primary region.
resource "azurerm_private_endpoint" "endpoint" {
  count = var.is_env_prod ? 2 : 1
  name                = "${azurerm_cosmosdb_account.account.name}-${count.index}" #This is where alphabetical order is introduced.
  location            = "${local.region_list[count.index]}"
  resource_group_name = azurerm_resource_group.resource_group.name
  subnet_id           = "${local.subnet_list[count.index]}"

  private_service_connection {
    is_manual_connection       = false
    name                       = "${azurerm_cosmosdb_account.account.name}-${local.region_list[count.index]}-connection"
    private_connection_resource_id = azurerm_cosmosdb_account.account.id
    subresource_names          = [var.account_kind]
  }
}


# Diagnostic settings - Log analytics

resource "azurerm_monitor_diagnostic_setting" "log_analytics" {
  
  provider = azurerm.loganalyticssub
  name               = "${azurerm_cosmosdb_account.account.name}-analytics"
  target_resource_id = "${azurerm_cosmosdb_account.account.id}"
  log_analytics_workspace_id = "${data.azurerm_log_analytics_workspace.ws.id}"

  dynamic "log" {
    for_each = data.azurerm_monitor_diagnostic_categories.diag_category.logs
    content {
    category = log.value
    enabled = true
    retention_policy {
      enabled = false
    }
    }
  }

  metric {
    category = "AllMetrics"

    retention_policy {
      enabled = false
    }
  }
}
```

*Error:*

```

Terraform v0.12.8
Initializing plugins and modules...
2020/02/24 16:42:50 [DEBUG] Using modified User-Agent: Terraform/0.12.8 TFE/975bcd7c07
azurerm_resource_group.resource_group: Creating...
azurerm_resource_group.resource_group: Creation complete after 2s [id=/subscriptionsXXXXXXXXXXXXXXXXXXXXX/resourceGroups/cosmosdb-privatelink-nonprod-c043ab60]
azurerm_cosmosdb_account.account: Creating...
azurerm_cosmosdb_account.account: Still creating... [10s elapsed]
azurerm_cosmosdb_account.account: Still creating... [20s elapsed]
azurerm_cosmosdb_account.account: Still creating... [30s elapsed]
azurerm_cosmosdb_account.account: Still creating... [40s elapsed]
azurerm_cosmosdb_account.account: Still creating... [50s elapsed]
azurerm_cosmosdb_account.account: Still creating... [1m0s elapsed]
azurerm_cosmosdb_account.account: Still creating... [1m10s elapsed]
azurerm_cosmosdb_account.account: Still creating... [1m20s elapsed]
azurerm_cosmosdb_account.account: Still creating... [1m30s elapsed]
azurerm_cosmosdb_account.account: Still creating... [1m40s elapsed]
azurerm_cosmosdb_account.account: Still creating... [1m50s elapsed]
azurerm_cosmosdb_account.account: Still creating... [2m0s elapsed]
azurerm_cosmosdb_account.account: Still creating... [2m10s elapsed]
azurerm_cosmosdb_account.account: Still creating... [2m20s elapsed]
azurerm_cosmosdb_account.account: Still creating... [2m30s elapsed]
azurerm_cosmosdb_account.account: Still creating... [2m40s elapsed]
azurerm_cosmosdb_account.account: Still creating... [2m50s elapsed]
azurerm_cosmosdb_account.account: Still creating... [3m0s elapsed]
azurerm_cosmosdb_account.account: Still creating... [3m10s elapsed]
azurerm_cosmosdb_account.account: Still creating... [3m20s elapsed]
azurerm_cosmosdb_account.account: Still creating... [3m30s elapsed]
azurerm_cosmosdb_account.account: Still creating... [3m40s elapsed]
azurerm_cosmosdb_account.account: Still creating... [3m50s elapsed]
azurerm_cosmosdb_account.account: Still creating... [4m0s elapsed]
azurerm_cosmosdb_account.account: Still creating... [4m10s elapsed]
azurerm_cosmosdb_account.account: Still creating... [4m20s elapsed]
azurerm_cosmosdb_account.account: Still creating... [4m30s elapsed]
azurerm_cosmosdb_account.account: Still creating... [4m40s elapsed]
azurerm_cosmosdb_account.account: Creation complete after 4m41s [id=/subscriptions/XXXXXXXXXXXXXXXXXXXXX/resourceGroups/cosmosdb-privatelink-nonprod-c043ab60/providers/Microsoft.DocumentDB/databaseAccounts/mongodb-cosmosdb-privatelink-c043ab60]
data.azurerm_monitor_diagnostic_categories.diag_category: Refreshing state...
azurerm_private_endpoint.endpoint[0]: Creating...
azurerm_monitor_diagnostic_setting.log_analytics: Creating...
azurerm_monitor_diagnostic_setting.log_analytics: Creation complete after 4s [id=/subscriptions/XXXXXXXXXXXXXXXXXXXXX/resourceGroups/cosmosdb-privatelink-nonprod-c043ab60/providers/Microsoft.DocumentDB/databaseAccounts/mongodb-cosmosdb-privatelink-c043ab60|mongodb-cosmosdb-privatelink-c043ab60-analytics]

Error: Error creating Private Endpoint "mongodb-cosmosdb-privatelink-c043ab60-0" (Resource Group "cosmosdb-privatelink-nonprod-c043ab60"): network.PrivateEndpointsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InternalServerError" Message="An error occurred." Details=[]

  on main.tf line 46, in resource "azurerm_private_endpoint" "endpoint":
  46: resource "azurerm_private_endpoint" "endpoint" {
```

This error ```400``` is happening consistently. I deleted RG and CosmosDB account created and tried to re-created, but again got ```400```.  It is such a generic error. 

Anyone has a clue what is happening?

**Comment by WodansSon** [comment]
@akamalov thank you for opening this issue, looking at your config file I don't see the configuration for your subnet, but would like to ask have you set the `enforce_private_link_endpoint_network_policies = true` on your subnet? If not this maybe the cause of your issue.

**Comment by akamalov** [comment]
Thanks so much @WodansSon for your prompt reply. I have in fact disabled ```enforce_private_link_endpoint_network_policies``` per recommendation - https://docs.microsoft.com/en-us/azure/private-link/disable-private-endpoint-network-policy which states the following:

```
When using the portal to create a private endpoint, this setting is automatically disabled as part 
of the create process. 
Deployment using other clients requires an additional step to change this
 setting. 
You can disable the setting using cloud shell from the Azure portal, or local installations of Azure PowerShell, Azure CLI, or use Azure Resource Manager templates.
```

Here is my subnet config:

```az network vnet subnet update -n sn_10.0.0.0_24 -g XXXXXXX_test_001 --vnet-name vn_XXXXXXX_test_001 --disable-private-endpoint-network-policies true
{
  "addressPrefix": "10.0.0.0/24",
  "addressPrefixes": null,
  "delegations": [],
  "etag": "W/\"15865106-5498-4e38-8f11-6abe6054072f\"",
  "id": "/subscriptions/XXXXXXXXXXXXXXX/resourceGroups/XXXXXXX_test_001/providers/Microsoft.Network/virtualNetworks/vn_XXXXXXX_test_001/subnets/sn_10.0.0.0_24",
  "ipConfigurationProfiles": null,
  "ipConfigurations": null,
  "name": "sn_10.0.0.0_24",
  "natGateway": null,
  "networkSecurityGroup": null,
  "privateEndpointNetworkPolicies": "Disabled",
  "privateEndpoints": null,
  "privateLinkServiceNetworkPolicies": "Enabled",
  "provisioningState": "Succeeded",
  "purpose": null,
  "resourceGroup": "XXXXXX_test_001",
  "resourceNavigationLinks": null,
  "routeTable": null,
  "serviceAssociationLinks": null,
  "serviceEndpointPolicies": null,
  "serviceEndpoints": null,
  "type": "Microsoft.Network/virtualNetworks/subnets"
}
```
Thanks yet again for looking into this.

Alex

**Comment by WodansSon** [comment]
What is the `var.account_kind`? I believe it should be `MongoDB` IIRC. I agree, that error returned from the RP is very, very generic.

**Comment by theindiandood** [comment]
Along with that. What is `var.primary_location`? As mentioned [here](https://azure.microsoft.com/en-us/updates/azure-private-endpoint-support-for-azure-cosmos-db-is-now-in-preview/) , currently there are only specific regions where you can have a private endpoint to cosmos.

**Comment by akamalov** [comment]
```var.primary_location``` is set to ```westus```

**Comment by theindiandood** [comment]
It looks like your `var.account_kind` is `mongodb`. Try changing it to `MongoDB`. It might be case sensitive.

**Comment by akamalov** [comment]
Thanks for your quick response, @theindiandood . The ```var.account_kind``` is actually ```MongoDB```. Here is my ```variables.tf``` file:

**File: variables.tf**

```
variable "is_env_prod" {
  type = bool
  default     = "true"
  description = "Set as true if environment is prod. Set false if environment is nonprod"
}

variable "primary_location" {
  type        = string
  default     = "westus"
  description = "Primary location to deploy the Cosmos DB resources"
}

variable "account_kind" {
  type        = string
  default     = "MongoDB"
  description = "Specifies the Kind of CosmosDB to create - possible values are GlobalDocumentDB and MongoDB. Defaults to GlobalDocumentDB. Changing this forces a new resource to be created."
}

variable "account_capabilities" {
  type        = string
  default     = "MongoDBv3.4"
  description = "The capabilities which should be enabled for this Cosmos DB account. Possible values are EnableAggregationPipeline, EnableCassandra, EnableGremlin, EnableTable, MongoDBv3.4, and mongoEnableDocLevelTTL."
}

variable "consistency_level" {
  default     = "BoundedStaleness"
  description = "The Consistency Level to use for this CosmosDB Account - can be either BoundedStaleness, Eventual, Session, Strong or ConsistentPrefix"
}

variable "secondary_location" {
  type        = string
  default     = "southcentralus"
  description = "Secondary region to deploy the Cosmos DB resources"
}
```

**Comment by WodansSon** [comment]
@akamalov I was able to repro your issue, this appears to be a bug in the Cosmos DB RP. I will be contacting the service team and opening a bug against their repo.

**Comment by prashantshetage** [comment]
@akamalov Hi. Were you able to resolve this?
I am getting the same error through Azure Portal as well while creating the private endpoint for CosmosDB, however it works fine for Storage account endpoint with the same configurations.
Looks like the problem is from Azure side

**Comment by akamalov** [comment]
No, I wasn't able to get it working...

**Comment by prashantshetage** [comment]
@akamalov 
I was able to create PE for CosmosDB

Private endpoint is supported for MongoDB version 3.6 only.
https://devblogs.microsoft.com/cosmosdb/azure-private-link-for-azure-cosmos-db-now-generally-available/

Terraform by default provisions CosmosDB-MongoDBv3.2, Hence Private Endpoint deployment fails.
 
Deployment succeeds after adding "EnableMongo" as capabilities.
This provisions CosmosDB-MongoDBv3.6 and support the private endpoint.
 
capabilities {
    name = "EnableMongo"
  }

**Comment by RichardFowles89** [comment]
I am getting this exact error, only that I am using GlobalDocumentDB and "SQL" as the subresource_names field for the endpoint rather than Mongo. Do I need to raise a new issue for this or can it go on this one?

**Comment by RichardFowles89** [comment]
I have resolved the problem I mentioned above. subresource_names appears to be very sensitive to casing. All lower and all upper didn't work. It has to be:

subresource_names = ["Sql"]

I hope this helps someone.

**Comment by gkaniouris** [comment]
@WodansSon Is there a public link for the issue you created? So that we can keep track of it.

**Comment by WodansSon** [comment]
@gkaniouris, I did not create and issue I pinged the service team directly.

**Comment by gkaniouris** [comment]
@WodansSon Ok. Do your recent updates mean that this will be fixed in 2.10.0?
